File: CPP/dnf_sort.cpp
// DNF Sort in CPP

#include<iostream>
using namespace std;

// dnfsort() function to sort the 3 numbers(0,1,2)
void dnfsort(int arr[], int n) {
    int low = 0, high = n-1, mid = 0;

    // while() until we reach the end of the function
    while(mid<=high) {

        // if the number is the smallest(0), we swap it to the start(low point)
        if(arr[mid]==0) {
            int temp = arr[mid];
            arr[mid] = arr[low];
            arr[low] = temp;
            low++;
            mid++;
        }

        // if the number is the largest(2), we swap it to the end(high point)
        else if(arr[mid]==2) {
            int temp = arr[mid];
            arr[mid] = arr[high];
            arr[high] = temp;
            high--;
        }

        // if the number is the middle one(1), we leave it and move on to next number
        else {
            mid++;
        }
    }
}

int main() {
    // Only 3 numbers can be taken in array to sort using DNF Sort
    int n;
    cin>>n;
    int arr[n];
    for(int i=0; i<n; i++) {
        cin>>arr[i];
    }

    // Sending the array to the dnfsort() function
    dnfsort(arr, n);

    // printing the sorted array
    for(int i=0; i<n; i++) {
        cout<<arr[i]<<" ";
    }
    cout<<endl;

    return 0;
}

/*
DNF SORT implementation in C++
Input: Size of array , Array Integers
Output: Returns the sorted array

----------------------------------------
Sample Input1:
6
0 2 2 1 1 0

Output:
0 0 1 1 2 2 

Sample Input2:
8
1 2 0 2 1 1 0 1

Output:
0 0 1 1 1 1 2 2

----------------------------------------

Time Complexity: O(n)

Space Complexity: O(1)

*/

File: Python/DNF_Sort.py
def dnfsort( a, size):
    low = 0 
    high = size - 1 
    mid = 0

    while mid <= high: 
        if a[mid] == 0: 
            a[low], a[mid] = a[mid], a[low]
            low = low + 1
            mid = mid + 1
        elif a[mid] == 1:
            mid = mid + 1 
        else:
            a[mid], a[high] = a[high], a[mid]
            high = high - 1
            
    return a 

ar= eval(input("Enter list:"))

size = len(ar) 

new= dnfsort( ar, size) 

print ("Array after segregation :", new)

"""
DNF SORT IMPLEMENTATION IN PYTHON

Please give input in this format only

Sample INPUT:

[0,1,2,0,1,2]  

OUTPUT: 

[0,0,1,1,2,2]


Time Complexity: O(N)

Space Complexity:O(1)
"""